### PR TITLE
Refactored generic graphs

### PR DESCRIPTION
**Pull request by drblallo**
Now generic graphs do not require crtp to be used.

### CODE REVIEW

File: include/revng/ADT/GenericGraph.h [hunk 1]
```diff
@@ -77,22 +87,62 @@ struct std::tuple_element<1, Edge<Node, EdgeLabel>> {
...
+template<typename Node,
+         typename EdgeLabel,
+         bool HasParent,
+         size_t SmallSize,
+         template<typename, typename, bool, size_t, typename>
+         class ForwardEdge,
+         typename MostDerivedType>
+struct ForwardNodeBaseTCalc {
+  static constexpr bool
+    NoMostDer = std::is_same_v<MostDerivedType, std::false_type>;
```
Comment (aleclearmind) [inline include/revng/ADT/GenericGraph.h:99]: Please do not use abbreviations, they're not readable.

File: include/revng/ADT/GenericGraph.h [hunk 2]
```diff
@@ -15,14 +17,22 @@ struct Empty {};
...
+  template<typename... Args>
+  explicit Parent(Args &&... args) :
+    BaseType(std::forward<Args>(args)...), Par(nullptr) {}
+
+  Parent(const Parent &) = default;
+  Parent(Parent &&) = default;
+
 private:
-  T *Parent;
+  T *Par;
```
Comment (aleclearmind) [inline include/revng/ADT/GenericGraph.h:31]: Why? Abbreviations are not readable.

File: include/revng/ADT/GenericGraph.h [hunk 3]
```diff
@@ -77,22 +87,62 @@ struct std::tuple_element<1, Edge<Node, EdgeLabel>> {
...
 } // namespace std
 
+template<typename Node,
+         typename EdgeLabel,
+         bool HasParent,
+         size_t SmallSize,
+         template<typename, typename, bool, size_t, typename>
+         class ForwardEdge,
+         typename MostDerivedType>
+struct ForwardNodeBaseTCalc {
```
Comment (aleclearmind) [inline include/revng/ADT/GenericGraph.h:97]: What's this? Please comment.

File: include/revng/ADT/GenericGraph.h [hunk 4]
```diff
@@ -77,22 +87,62 @@ struct std::tuple_element<1, Edge<Node, EdgeLabel>> {
...
-template<typename Inheritor,
+template<typename Node,
          typename EdgeLabel = Empty,
          bool HasParent = true,
-         size_t SmallSize = 2>
-class ForwardNode : public std::conditional_t<HasParent,
-                                              Parent<GenericGraph<Inheritor>>,
-                                              Empty> {
+         size_t SmallSize = 2,
+         typename MostDerivedType = std::false_type>
```
Comment (aleclearmind) [inline include/revng/ADT/GenericGraph.h:116]: `s/MostDerived/Final/`?

File: include/revng/ADT/GenericGraph.h (hunk 1 above)
Comment (drblallo) [reply include/revng/ADT/GenericGraph.h:99]: i had to or clang tidy would have broken the line

File: include/revng/ADT/GenericGraph.h (hunk 2 above)
Comment (drblallo) [reply include/revng/ADT/GenericGraph.h:31]: when a costructor is added with the same name as the class, no field can have the same name as the calss

Comment (aleclearmind) [reply include/revng/ADT/GenericGraph.h:31]: What about `TheParent`?

File: include/revng/ADT/GenericGraph.h (hunk 1 above)
Comment (aleclearmind) [reply include/revng/ADT/GenericGraph.h:99]: I suggest `IsFinal` and then use `!IsFinal`

File: include/revng/ADT/GenericGraph.h (hunk 3 above)
Comment (aleclearmind) [reply include/revng/ADT/GenericGraph.h:97]: Can you please hide this in a `detail` namespace?

File: include/revng/ADT/GenericGraph.h (hunk 1 above)
Comment (drblallo) [reply include/revng/ADT/GenericGraph.h]: renamed in NoDerivation

File: include/revng/ADT/GenericGraph.h [hunk 5]
```diff
@@ -166,43 +223,85 @@ class ForwardNode : public std::conditional_t<HasParent,
...
+                  bool,
+                  size_t>
+         class BidirectionalNode> // At this step BidirectionalEdge has not been
+                                  // declared yet, thus we accept a template
+                                  // parameter that has the same signature as
+                                  // BidirecationEdge that will be declared
+                                  // later. This allows us to use it as if it
+                                  // was delcared, provided that only the real
+                                  // BidirecationalEdge is used as this
+                                  // argument.
```
Comment (aleclearmind) [inline include/revng/ADT/GenericGraph.h:245]: These comments are too long for being inline. Put them in a `/// \note:` comment above.

**Comment by aleclearmind** [comment]
Merged in 8ba9930638446381303011d68fa4aacd0bee6a4e.
Thanks!
A couple of fixes I did:
* Change the commit message to a more descriptive one
* Rewrap the `\note` comment to 80 cols (do you have a tool that does that for you?)

Thanks again.

**Pull request closed without merging**